src/platform_impl/web/monitor.rs

```diff
@@ -1,53 +1,798 @@
+                    },
+                };
+
+                // Notifying `Future`s is not dependant on the lifetime of the runner,
+                // because they can outlive it.
+                if let Some(runner) = runner.upgrade() {
+                    let state = if let Some(screen_details) = screen_details {
+                        State::Detailed(screen_details)
+                    } else {
+                        // If permission is denied we listen for changes so we can catch external
+                        // permission granting.
+                        let handle = EventListenerHandle::new(
+                            permission.clone(),
+                            "change",
+                            Closure::new({
+                                let runner = runner.weak();
+                                let permission = permission.clone();
+                                move || {
+                                    if let PermissionState::Granted = permission.state() {
+                                        let future = JsFuture::from(window.screen_details());
+
+                                        let runner = runner.clone();
+                                        wasm_bindgen_futures::spawn_local(async move {
+                                            let screen_details = match future.await {
+                                                Ok(screen_details) => {
+                                                    screen_details.unchecked_into()
+                                                },
+                                                Err(error) => unreachable_error(
+                                                    &error,
+                                                    "getting screen details failed even though \
+                                                     permission was granted",
+                                                ),
+                                            };
+
+                                            if let Some(runner) = runner.upgrade() {
+                                                // We drop the event listener handle here, which
+                                                // doesn't drop it while we are running it, because
+                                                // we are in a `spawn_local()` context.
+                                                *runner.monitor().state.borrow_mut() =
+                                                    State::Detailed(screen_details);
+                                            }
+                                        });
+                                    }
+                                }
+                            }),
+                        );
+                        State::Permission { permission, _handle: handle }
+                    };
+
+                    *runner.monitor().state.borrow_mut() = state;
+                    runner.start_delayed();
```

Comment:
All of that is just a massive indention level and the function gets really long, should probably split, use early returns, etc to maintain the readability.

Reply:
Refactored.